### PR TITLE
feat: add image upload

### PR DESCRIPTION
**Pull request by SinghHarman286**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=1b2a2134dfa84cb1b461fea7a58a70a8)

## Implementation description
* added image upload middleware in `POST /camp` route, where image and camp data should be sent in `form-data`. `form-data` contains 2 keys: file containing the image that needs to be uploaded, and body containing camp data that needs to be saved. When image is successfully uploaded, a key: `fileName` containing image name gets stored in db along with the camp data
* "file" key in form-data is optional i.e. uploading an image in `POST /camp` route is optional. If the image is not provided, then `fileName` key will not be created in db
* max image upload size is 5MB

## Steps to test
1. In order to test the changes, go to postman->body->form-data. Create 2 keys: file containing the image that needs to be uploaded, and body containing the camp data that needs to be stored in db. You can use the below sample body input:
`{"active": true, "ageLower": 15, "ageUpper": 30, "name": "test-camp-image", "description": "image-test-description", "location": "canada", "capacity": 20,  "fee": 25,  "formQuestions": [{"question": "Q1", "type": "Text", "required": true, "description": "question description"}], "campSessions": []}`

Screenshot 2022-04-13 at 4 28 25 PM

In order to verify if correct image is uploaded, you can refer to this [link](https://console.firebase.google.com/u/6/project/focus-on-nature-6c0b1/storage/focus-on-nature-6c0b1.appspot.com/files)

## What should reviewers focus on?
* does `fileName` key needs to be created if there is no image to be uploaded?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by amanda-guo — COMMENTED** [review]
uploading an image works as expected. Also tested non-jpeg/png and a pic > 5 MB and this works as well (throws errors and no entries created in DB). Merge conflicts do need to be resolved to make sure the image upload is just for `camps`, so I can take a look at this again after that's done!

**Review by amanda-guo — COMMENTED** [review]
One thing...not sure if this was in the scope of the ticket, but are there supposed to be changes to the edit camp API for the images or is that a separate ticket?

**Comment by SinghHarman286** [comment]
> One thing...not sure if this was in the scope of the ticket, but are there supposed to be changes to the edit camp API for the images or is that a separate ticket?

Currently, there is no edit camp API and I believe a separate ticket needs to be made for that. cc @annechung

**Review by amanda-guo — APPROVED** [review]
works as expected! also i think we probably don't need to create a `fileName` key if there is no image uploaded and probably just add one if there is an image added later

**Comment by SinghHarman286** [comment]
> works as expected! also i think we probably don't need to create a `fileName` key if there is no image uploaded and probably just add one if there is an image added later

I think that's the way it is happening currently. If no image is provided, then fileName key is not created: [link](https://github.com/uwblueprint/focus-on-nature/blob/00983dc48b40b52fe5a7010d2af2d43f75904f1e/backend/typescript/services/implementations/campService.ts#L181)

File: backend/typescript/middlewares/validators/campValidators.ts [hunk 1]
```diff
@@ -92,11 +100,17 @@ export const createCampDtoValidator = async (
...
   }
-  if (req.body.campers) {
+  if (body.campers) {
     return res.status(400).send("campers should be empty");
   }
-  if (req.body.waitlist) {
+  if (body.waitlist) {
     return res.status(400).send("waitlist should be empty");
   }
+  if (req.file && !validateImageType(req.file.mimetype)) {
```
Comment (annechung) [inline backend/typescript/middlewares/validators/campValidators.ts:109]: super nitpicky but: can we change the variable name to something a bit more descriptive (e.g. campPhoto?)

File: .gitignore [hunk 2]
```diff
@@ -9,3 +9,4 @@
 .vscode
 **/*.cache
 **/*.egg-info
+**/uploads/
```
Comment (annechung) [inline .gitignore:12]: what does this directory store? :o

File: backend/typescript/rest/campRoutes.ts [hunk 3]
```diff
@@ -19,25 +29,35 @@ campRouter.get("/", async (req, res) => {
...
+        location: body.location,
+        capacity: body.capacity,
+        fee: body.fee,
+        formQuestions: body.formQuestions,
+        campSessions: body.campSessions,
+        filePath: req.file?.path,
+        fileContentType: req.file?.mimetype,
+      });
+      if (req.file?.path) {
+        fs.unlinkSync(req.file.path);
```
Comment (annechung) [inline backend/typescript/rest/campRoutes.ts:53]: what is this doing?

**Review by annechung — COMMENTED** [review]
tested and the image upload works properly, nice work :) 
just confirming, the file name is the identifier for the image that's uploaded? (and we can use the file name to get the url of the image?)

File: backend/typescript/rest/campRoutes.ts (hunk 3 above)
Comment (SinghHarman286) [reply backend/typescript/rest/campRoutes.ts:53]: Each image thats uploaded is temporarily stored in "upload" directory, so this `fs.unlinkSync(req.file.path)` deletes the file from that directory. For example, req.file.path is something like 'uploads/152280faf4e59f64175f8a7186a48eab', so `fs.unlinkSync(req.file.path)` will just remove `152280faf4e59f64175f8a7186a48eab` from the upload directory

File: backend/typescript/middlewares/validators/campValidators.ts (hunk 1 above)
Comment (SinghHarman286) [reply backend/typescript/middlewares/validators/campValidators.ts:109]: oh you mean like something like validateCampPhotoType?

File: backend/typescript/middlewares/validators/campValidators.ts [hunk 4]
```diff
@@ -15,39 +20,42 @@ export const createCampDtoValidator = async (
   res: Response,
   next: NextFunction,
 ) => {
-  if (!validatePrimitive(req.body.name, "string")) {
+  let body;
+  try {
+    body = JSON.parse(req.body.body);
```
[the remark this answers is not in the document]
Comment (SinghHarman286) [reply backend/typescript/middlewares/validators/campValidators.ts:25]: I think formData can support either a string or a blob type. If we add any fields to formdata, then its value will be a string, so I guess we'll have write another logic to parse those value in their correct data type which I think might result in complex code and more errors? https://developer.mozilla.org/en-US/docs/Web/API/FormData/append

File: .gitignore (hunk 2 above)
Comment (SinghHarman286) [reply .gitignore:12]: it stores the files that does not get uploaded because of invalid content type

Comment (SinghHarman286) [reply .gitignore:12]: As I am thinking a bit more about this, I think I should use `fs.unlinkSync` in the validator class as well just before we are returning invalid content type error response. That way, the files with invalid content type will not get stored in the uploads directory

File: backend/typescript/rest/campRoutes.ts (hunk 3 above)
Comment (annechung) [reply backend/typescript/rest/campRoutes.ts:53]: ah ok gotcha

File: backend/typescript/middlewares/validators/campValidators.ts (hunk 1 above)
Comment (annechung) [reply backend/typescript/middlewares/validators/campValidators.ts:109]: i meant `file`!

File: .gitignore (hunk 2 above)
Comment (annechung) [reply .gitignore:12]: got it. removing invalid files sounds good to me, and yep should ignore uploads

File: backend/typescript/middlewares/validators/campValidators.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (annechung) [reply backend/typescript/middlewares/validators/campValidators.ts:25]: ah I see - we can keep it like this then! can we rename `body` to `data` though? `body.body` just looks funny to me. up to you, not a huge deal either way

File: backend/typescript/middlewares/validators/campValidators.ts (hunk 1 above)
Comment (SinghHarman286) [reply backend/typescript/middlewares/validators/campValidators.ts:109]: adding custom properties to `Request` object would result in typescript throwing an [error](https://stackoverflow.com/questions/58957228/property-does-not-exist-on-type-requestparamsdictionary). In order to fix it, we'll need to add additional changes to `Request`'s type definition which could make the code more complex. Thus, sticking with property name `file`.

**Comment by SinghHarman286** [comment]
> tested and the image upload works properly, nice work :) just confirming, the file name is the identifier for the image that's uploaded? (and we can use the file name to get the url of the image?)

Yes, file name is the identifier and we'll use something like getFile(fileName: string) in order to retrieve the file -> https://uwblueprint.github.io/starter-code-v2/docs/file-storage/typescript-file-storage-service.html#getfile

**Review by annechung — APPROVED** [review]
looks good! I think there are cases where camp creation fails but the image will still be uploaded which we can look into separately, but this is definitely a strong start. awesome work :D

**Pull request merged**